currencies/rate_converter.go

```diff
@@ -0,0 +1,205 @@
+func (rc *RateConverter) fetch() (*Rates, error) {
+	request, err := http.NewRequest("GET", rc.syncSourceURL, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), 3*time.Second)
+	defer cancel()
+	request = request.WithContext(ctx)
+
+	response, err := rc.httpClient.Do(request)
+	if err != nil {
+		return nil, err
+	}
+
+	defer response.Body.Close()
+
+	bytesJSON, err := ioutil.ReadAll(response.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	updatedRates := &Rates{}
+	err = json.Unmarshal(bytesJSON, &updatedRates)
+	if err != nil {
+		return nil, err
+	}
+
+	return updatedRates, err
+}
+
+// updateRates allows to update the internal currencies rates map with the one provided
+func (rc *RateConverter) updateRates(updatedRates *Rates) {
+	rc.rates.Store(updatedRates)
+	rc.lastUpdated.Store(time.Now())
+}
+
+// Update updates the internal currencies rates from remote sources
+func (rc *RateConverter) Update() (*Rates, error) {
+	rates, err := rc.fetch()
+	if err == nil {
+		rc.updateRates(rates)
+	}
+
+	return rates, err
+}
+
+// startPeriodicFetching starts the periodic fetching at the given interval
+// triggers a first fetch when called before the first tick happen in order to initialize currencies rates map
+// returns a chan in which the number of data updates everytime a new update was done
+func (rc *RateConverter) startPeriodicFetching() error {
+
+	rc.ticker = time.NewTicker(rc.fetchingInterval)
+	updatesTicksCount := 0
+
+	go func() {
+		for {
+			select {
+			case <-rc.ticker.C:
+				go func() {
+					rates, err := rc.fetch()
+					if err == nil {
+						rc.updates <- rates
+					}
+					// TODO: Handle retry strategy
+				}()
+			case rates := <-rc.updates:
+				go func() {
+					rc.updateRates(rates)
+					updatesTicksCount++
+					if rc.updateNotifier != nil {
+						rc.updateNotifier <- updatesTicksCount
+					}
+				}()
+			case <-rc.done:
+				if rc.ticker != nil {
+					rc.ticker.Stop()
+					rc.ticker = nil
+				}
+				return
+			default:
+			}
+		}
+	}()
+
+	return nil
+}
+
+// StopPeriodicFetching stops the periodic fetching while keeping the latest currencies rates map
+func (rc *RateConverter) StopPeriodicFetching() {
+	rc.done <- true
+}
+
+// LastUpdated returns time when currencies rates were updated
+func (rc *RateConverter) LastUpdated() time.Time {
+	if lastUpdated := rc.lastUpdated.Load(); lastUpdated != nil {
+		return lastUpdated.(time.Time)
+	}
+	return time.Time{}
+}
+
+// Rates returns current rates
+func (rc *RateConverter) Rates() *Rates {
+	if rates := rc.rates.Load(); rates != nil {
+		return rates.(*Rates)
+	}
+	return nil
+}
+
+// GetRate returns the conversion rate between two currencies
+// returns an error in case the conversion rate between the two given currencies is not in the currencies rates map
+func (rc *RateConverter) GetRate(from string, to string) (float64, error) {
```

Comment:
I am thinking we may want to hang this function on the `Rates` struct rather than the `RateConverter` object. This way, an auction can call to get the `Rates` once, at the expense of one atomic transaction, rather than invoking the atomic machinery for each conversion call. This would also avoid the race condition where the exchange rates could change half way through an auction, not that this would be a serious bug.

Reply:
Good point, just move this function to `Rates` struct.